# 06. Ensemble — 複数モデルのブレンド

README の工程 **⑥ Ensemble**。本番スクリプトは `src/06_ensemble_hill_climbing.py`。
このノートブックは同じ処理を分解して、**なぜこの構成になったか**を追えるようにしたもの。

## ここまでのフローまとめ

| 段階 | CV | Public LB |
|---|---|---|
| ベースライン(CatBoost単体) | 0.94156 | 0.94170 |
| FE(厳密値TE ほか) | 0.94589 | — |
| HPO(列サブサンプリング) | 0.94610 | — |
| アンサンブル(3モデル) | 0.94623 | **0.94645** |
| **(New)アンサンブル(現在の構成。2026-09-28)** | **0.946264** | 0.94642 |

単体の最良(LightGBM 0.94610)に対し、ブレンドで **+0.00013**。
数字は小さいが、**2026-09-23 時点で上位15%のラインが LB 0.94641、上位10%が 0.94647** と
0.0001 未満の刻みで順位が動く世界なので、十分に意味のある差になる。

## なぜブレンドが効くのか

モデルごとに予測の間違え方が違うため、平均すると誤差が打ち消し合う。
したがって**単体の強さより「予測の間違え方が違うこと(非相関性)」が重要**になる。

実際このコンペでも、単体で最弱(0.94462)だった LightGBM のバリアントが、
非相関だったために重み 25% で採用された局面があった(**diversity beats strength**)。

In [1]:
import os, sys
if os.path.basename(os.getcwd()) == "notebooks":
    os.chdir("..")
sys.path.insert(0, os.path.abspath("src"))
# 絶対パスは出力しない(公開リポジトリに個人のディレクトリ構成を残さないため)
print("oof/ を検出:", os.path.isdir("oof"))

oof/ を検出: True


In [2]:
import glob, importlib, itertools
import numpy as np
import pandas as pd
from scipy.stats import rankdata, spearmanr
from sklearn.metrics import roc_auc_score

train = pd.read_csv("data/train.csv")
test = pd.read_csv("data/test.csv")
y = (train["Will_Buy_EV"] == "Yes").astype(int).values
print(f"train {len(train):,} / test {len(test):,}")

train 668,665 / test 286,571


## 候補の読み込み

`oof/` にある **OOF 予測(学習に使わなかった行への予測)** と、対応する test 予測を全部拾う。

各モデルは fold 分割を `StratifiedKFold(5, shuffle=True, random_state=42)` で**完全に揃えて**いる。
これが揃っていないと OOF 同士を足すことも、後の DeLong 検定もできない。

### なぜ rank に変換するのか

モデルごとに出力する確率のスケールが違う(あるモデルは 0.2 前後に集中、別のモデルは広く散る)。
そのまま平均すると散らばりの大きいモデルの影響が強く出てしまう。
**順位に変換してから平均**すれば、各モデルを対等に扱える。AUC は順位だけで決まる指標なので、
この変換で情報は失われない。

In [3]:
cands = {}
for f in sorted(glob.glob("oof/oof_*.npy")):
    name = os.path.basename(f).replace("oof_", "").replace(".npy", "")
    pred_path = f"oof/pred_{name}.npy"
    if os.path.exists(pred_path):
        cands[name] = (rankdata(np.load(f)) / len(y),          # OOF(順位に変換)
                       rankdata(np.load(pred_path)) / len(test))  # test(同上)

tbl = pd.DataFrame(
    [(n, roc_auc_score(y, o)) for n, (o, _) in cands.items()],
    columns=["候補", "OOF AUC"]
).sort_values("OOF AUC", ascending=False).reset_index(drop=True)
print(f"候補 {len(cands)} 本")
tbl.head(10)

候補 17 本


,候補,OOF AUC
0,lgbm,0.946109
1,lgbm_shallow,0.946095
2,xgb,0.946087
3,xgb_shallow,0.946077
4,realmlp,0.946025
5,catboost,0.945924
6,xgb_pre_shallow,0.945913
7,lgbm_pre_shallow,0.945872
8,lgbm_tte_dig,0.945872
9,realmlp_e6,0.945849


## 主力4モデルの相関

**ここが最も重要な図。** 相関が高いほど「同じ予測の間違え方」をしており、足しても伸びない。

In [4]:
main = [m for m in ["lgbm", "xgb", "catboost", "realmlp"] if m in cands]
rows = []
for a, b in itertools.combinations(main, 2):
    rows.append([a, b, spearmanr(cands[a][0], cands[b][0]).statistic])
corr = pd.DataFrame(rows, columns=["A", "B", "順位相関"]).sort_values("順位相関")
corr

,A,B,順位相関
2,lgbm,realmlp,0.994733
4,xgb,realmlp,0.995265
5,catboost,realmlp,0.995657
1,lgbm,catboost,0.995750
3,xgb,catboost,0.996353
0,lgbm,xgb,0.997823


GBDT 同士(LightGBM / XGBoost / CatBoost)は **0.993〜0.999** と極めて似ている。
特に LightGBM × XGBoost は、両方に同じ「列サブサンプリング + 深さ5」を入れた結果 **0.9986** まで同質化した。

一方 **RealMLP だけが 0.994 前後と低い**。ニューラルネットで仕組みが根本的に違うため、
ここが多様性の供給源になっている。現在の重み 1/3 はこの非相関性によるもので、
単体スコア(0.94603。LightGBM・XGBoost より低い)で選ばれているわけではない。

## hill climbing(貪欲法)

空の状態から始めて、**「今の平均に足すと最も良くなる候補」を1本ずつ選ぶ**だけの単純な方法。
同じ候補を複数回選べるようにしてあり、**選ばれた回数がそのまま重み**になる。

Nelder-Mead などで重みを連続的に最適化する方法もあるが、候補が20本以上あると
過学習しやすい。貪欲法は選択肢が離散的なぶん、過学習しにくい。

In [5]:
selected, cur_sum, best_auc = [], np.zeros(len(y)), 0.0
names = list(cands)

for step in range(15):
    scores = {n: roc_auc_score(y, (cur_sum + cands[n][0]) / (len(selected) + 1)) for n in names}
    pick = max(scores, key=scores.get)
    if scores[pick] <= best_auc + 1e-7:   # 改善が止まったら終了
        break
    best_auc = scores[pick]
    selected.append(pick)
    cur_sum = cur_sum + cands[pick][0]
    print(f"step {step+1}: +{pick:20s} -> {best_auc:.5f}")

weights = {n: selected.count(n) / len(selected) for n in set(selected)}
print(f"\n最終 OOF AUC: {best_auc:.5f}")
print("重み:", {k: round(v, 4) for k, v in sorted(weights.items(), key=lambda kv: -kv[1])})

step 1: +lgbm                 -> 0.94611


step 2: +realmlp              -> 0.94625


step 3: +xgb                  -> 0.94626



最終 OOF AUC: 0.94626
重み: {'lgbm': 0.3333, 'realmlp': 0.3333, 'xgb': 0.3333}


## ⚠ ここで終わらせない — 工程⑦(DeLong 検定)が必要な理由

**貪欲法は OOF 上の偶然を拾う。** 候補を20本以上並べて「一番良くなるもの」を選び続ければ、
たまたま OOF に合っただけの組み合わせが選ばれうる。

実際にそれで失敗した記録がある。

| 構成 | CV | z値 | Public LB |
|---|---|---|---|
| 3モデル | 0.946234 | — | **0.94645** |
| 5点(貪欲法が「良い」と判断) | 0.946243(**+0.000009**) | +1.57(有意でない) | **0.94643**(-0.00002) |

CV では良くなったのに、**LB では逆に下がった**。DeLong 検定は事前に「有意でない」と判定していた。

`src/07_compare_predictions.py` はこの判定を行う。**スコアを作るスクリプトではなく、
採用するかどうかを決める「関門」**である。

In [6]:
cmp = importlib.import_module("07_compare_predictions")

ens = cur_sum / len(selected)                       # 今回のアンサンブル
best_single = max(main, key=lambda m: roc_auc_score(y, cands[m][0]))

a0, a1, diff, se, z, p = cmp.paired_test(y, cands[best_single][0], ens)
print(f"単体最良 ({best_single}): {a0:.6f}")
print(f"アンサンブル          : {a1:.6f}")
print(f"差分 {diff:+.6f}  SE {se:.6f}  z {z:+.2f}  p {p:.2e}")
print("→ 採用" if (diff >= 8e-5 and z >= 3) else "→ 誤差(採用しない)")

単体最良 (lgbm): 0.946109
アンサンブル          : 0.946264
差分 +0.000155  SE 0.000015  z +10.15  p 3.21e-24
→ 採用


## 提出ファイルの作成

OOF で決めた重みを、そのまま test 予測に適用する。

In [7]:
test_pred = sum(w * cands[n][1] for n, w in weights.items())
sub = pd.DataFrame({"id": test["id"], "Will_Buy_EV": test_pred})

# 本番の submit/submission_hillclimb.csv を上書きしないよう別名で保存する
sub.to_csv("submit/submission_from_notebook.csv", index=False)
print(sub.shape)
sub.head()

(286571, 2)


,id,Will_Buy_EV
0,668665,0.557349
1,668666,0.493059
2,668667,0.293396
3,668668,0.251429
4,668669,0.554315


## 提出履歴

| 提出 | 構成 | CV | Public LB | 順位相当 |
|---|---|---|---|---|
| CatBoost 単体 | — | 0.94156 | 0.94170 | 1637位 |
| 4モデル | RealMLP .4 / CatBoost .2 / LGBM .2 / XGB .2 | 0.94618 | 0.94636 | 523位 |
| **3モデル** | **LGBM 1/3 / XGB 1/3 / RealMLP 1/3** | **0.94623** | **0.94645** | **293位** |
| 5点 | 上記 + 2バリアント | 0.946243 | 0.94643 | (下がった) |
| 3モデル(RealMLP の比を削除) | LGBM 1/3 / XGB 1/3 / RealMLP 1/3 | 0.946230 | 0.94644 | (誤差) |
| 3モデル(さらに通勤距離 /5 を削除) | LGBM 1/3 / XGB 1/3 / RealMLP 1/3 | 0.946227 | 0.94644 | (誤差) |
| 3モデル(GBDT の重複削除・エンコーディングの絞り込み) | LGBM 1/3 / XGB 1/3 / RealMLP 1/3 | 0.946232 | 0.94642 | (誤差) |
| D: 上記 + LightGBM・RealMLP に自宅充電 × スタンド数 | LGBM 1/3 / XGB 1/3 / RealMLP 1/3 | 0.946245 | 0.94640 | (置き換え) |
| **現在の構成: D + RealMLP に年収の Target Encoding(train だけで作る)** | LGBM 1/3 / XGB 1/3 / RealMLP 1/3 | **0.946264** | 0.94642 | **最終提出** |

> 順位は **2026-09-23 時点の暫定値**(2,732チーム)の LB で、そのスコアが今何位に相当するかを
> 示したもの。提出当時の順位ではない。**09-23 時点の実際の順位は 320位**(09-28 時点では 459位 / 3,295チーム)(0.94645 に 32チームが並んでおり、
> Kaggle は同点を提出時刻順に並べるため)。

> 2026-09-27 に、RealMLP から効果のなかった特徴量(通勤距離 ÷ 年齢、通勤距離 /5 のキー)を削除した。3モデルの CV は 0.946234 → 0.946230 → 0.946227 で、それぞれの差は誤差(z=-0.88 / -0.59)。提出した Public LB も 0.94645 → 0.94644 → 0.94644 で、同じく誤差の範囲だった(Public は test の 20% だけで計算されるため、0.00001 の差は区別できない)。

> **最終提出(2本)**: CV が最高の**現在の構成**(CV 0.946264 / Public 0.94642。2026-09-28 に D から置き換え)と、Public が最高の**元の構成**(CV 0.946234 / Public 0.94645)。CV の差はどれも有意でないので、Private に近い CV の最高と、Public の最高を1本ずつ選んだ。

**得られた教訓**

1. **単体を強くすることと多様性はトレードオフ。** 4モデル全てを同じ最強構成に揃えたとき、
   同質化してアンサンブルの上積みが +0.00002 まで縮んだ
2. **弱くても非相関なら採用される。** 単体最弱のバリアントが重み25%を得た局面があった
3. **CV の非有意な改善は LB で再現しない。** 必ず DeLong 検定を通してから提出する
4. CatBoost は単体では健闘していた(0.94592)が、**他のGBDTと同質だったため最終的に重み0**になった